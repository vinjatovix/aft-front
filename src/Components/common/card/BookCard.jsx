import React from "react";
import PropTypes from "prop-types";
import { NavLink } from "react-router-dom";
import { ListItem } from "../ListItem";

export const BookCard = ({ data, isAdmin, actions, token }) => {
  const defaultCoverBook =
    "https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg";
  return (
    <div className="flex-wrapper column">
      <ul
        className="book"
        onClick={() => {
          actions.detail();
        }}
      >
        {<img src={data.img ? data.img : defaultCoverBook} alt={data.name} />}
        <ListItem name="Obra" children={data.name} />
        <ListItem name="Autor" className="small" children={data.author} />
      </ul>
      <NavLink to={`/characters/book/${data._id}`}>
        <button data-testid="char-button" className="bg-blue">
          🎭 Personajes
        </button>
      </NavLink>
      {isAdmin && (
        <div className="control-panel-mini">
          <button
            className="bg-yellow"
            data-testid="edit-button"
            onClick={() => {
              actions.edit();
            }}
          >
            📝 editar
          </button>
          <button
            data-testid="delete-button"
            className="bg-red"
            onClick={() => {
              actions.delete(token, data._id);
            }}
          >
            🗑️ eliminar
          </button>
        </div>
      )}
    </div>
  );
};

BookCard.propTypes = {
  data: PropTypes.shape({
    name: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    img: PropTypes.string,
  }).isRequired,
  isAdmin: PropTypes.bool,
  actions: PropTypes.object.isRequired,
  token: PropTypes.string.isRequired,
};

BookCard.defaultProps = {
  isAdmin: false,
};
