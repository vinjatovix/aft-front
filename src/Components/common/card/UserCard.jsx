import React from "react";
import PropTypes from "prop-types";

import { ListItem } from "../ListItem";
import { UserRole } from "../../ui/UserRole";

export const UserCard = ({ data }) => {
  return (
    <ul className="user-list">
      <ListItem name="Nombre" children={data.username} />
      <ListItem name="Role" children={UserRole(data.roles)} />
    </ul>
  );
};

UserCard.propTypes = {
  data: PropTypes.shape({
    username: PropTypes.string.isRequired,
    roles: PropTypes.arrayOf(PropTypes.string).isRequired,
  }).isRequired,
};
