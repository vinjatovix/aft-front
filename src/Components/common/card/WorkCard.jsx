import React from "react";
import PropTypes from "prop-types";
import { ListItem } from "../ListItem";

export const WorkCard = ({ data }) => {
  return (
    <ul className="work-list">
      <ListItem
        name="Intérprete"
        children={<p>by: {data.metadata.createdBy}</p>}
        liKey={data.metadata.createdBy}
      />
      <ListItem
        name="Obra"
        children={<p>Obra: {data.character.book.name}</p>}
        liKey={data.character.book._id}
      />
      <ListItem
        name="Personaje"
        children={<p>Personaje: {data.character.name}</p>}
        liKey={data.character._id}
      />
      <ListItem
        name="Escena"
        children={<p>Escena: {data.scene.name}</p>}
        liKey={data.scene._id}
      />
      <ListItem
        name="Descripción"
        children={<p>Descripción: {data.description}</p>}
      />
    </ul>
  );
};

WorkCard.propTypes = {
  data: PropTypes.object.isRequired,
};
