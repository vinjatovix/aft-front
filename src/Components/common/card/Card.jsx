import React from "react";
import PropTypes from "prop-types";
import { UserCard } from "./UserCard";

const Card = ({ data, type, actions, isAdmin, token }) => {
  return (
    <div className="card" onClick={() => actions.setDataDetail(data)}>
      {type === "user" && <UserCard data={data} />}
    </div>
  );
};

Card.propTypes = {
  type: PropTypes.string.isRequired,
  data: PropTypes.object.isRequired,
  actions: PropTypes.object,
  isAdmin: PropTypes.bool,
  token: PropTypes.string,
};

Card.defaultProps = {
  isAdmin: false,
};

export default Card;
