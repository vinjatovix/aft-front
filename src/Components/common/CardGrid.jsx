import React from "react";
import PropTypes from "prop-types";
import Card from "./card/Card";

export const CardGrid = ({ data, actions, isAdmin, type, token }) => {
  return (
    <div className="card-grid">
      {data.map((item) => (
        <Card
          type={type}
          key={item._id}
          data={item}
          isAdmin={isAdmin}
          actions={actions}
          token={token}
        />
      ))}
    </div>
  );
};

CardGrid.propTypes = {
  type: PropTypes.string.isRequired,
  data: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired,
  isAdmin: PropTypes.bool,
  token: PropTypes.string,
};

CardGrid.defaultProps = {
  isAdmin: false,
};
