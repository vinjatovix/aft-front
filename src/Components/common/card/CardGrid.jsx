import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { modalActionsType, modalsStateType } from "../../../PropTypes";
import { api } from "../../../http/api";

import { Card } from "./Card";
import { ContentHeader } from "../content/ContentHeader";

import { useFetch } from "../../../hooks/useFetch";

export const CardGrid = ({
  actions,
  auth: { token },
  isEditor,
  modals,
  type,
}) => {
  const [{ data, loading, error }, reFetch] = useFetch(
    api[type].getAll.path,
    api[type].getAll.method,
    { token }
  );

  const title = type === "work" ? "Trabajos:" : "Obras:";
  const count = data ? data.length : 0;

  useEffect(() => {
    token && reFetch();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [modals.lastRefresh, token]);

  return (
    <>
      <ContentHeader count={count} title={title} />
      <div className="card-grid">
        {data &&
          data.map((item) => (
            <Card
              actions={actions}
              data={item}
              isEditor={isEditor}
              key={item._id}
              token={token}
              type={type}
            />
          ))}
        {error && <div>{error}</div>}
        {loading && <div>Cargando...</div>}
      </div>
    </>
  );
};

CardGrid.propTypes = {
  actions: modalActionsType,
  auth: PropTypes.shape({
    token: PropTypes.string.isRequired,
  }),
  isEditor: PropTypes.bool,
  modals: modalsStateType,
  type: PropTypes.string.isRequired,
};

CardGrid.defaultProps = {
  isEditor: false,
};
