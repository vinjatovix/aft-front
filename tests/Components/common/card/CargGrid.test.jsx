import { render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { CardGrid } from "../../../../src/Components/common/card/CardGrid";
import mockActions from "../../../fixtures/mockActions";
import { mockModals } from "../../../fixtures/mockModals";
import mockWorks from "../../../fixtures/mockWorks";
import mockBooks from "../../../fixtures/mockBooks";

jest.mock("../../../../src/hooks/useFetch", () => ({
  useFetch: (path) => [
    {
      data: /work/.test(path) ? mockWorks : mockBooks,
      loading: false,
      error: null,
    },
    jest.fn(),
  ],
}));

jest.mock("../../../../src/helpers/getApiHost", () => ({
  getApiHost: () => "host",
}));

const auth = {
  token: "token",
  user: {
    _id: "5e9f8f8f8f8f8f8f8f8f8f8",
    username: "some",
    roles: ["aft.user"],
  },
};
describe("CardGrid Component", () => {
  it("should render 2 work cards", () => {
    const { container } = render(
      <Router>
        <CardGrid
          type="work"
          actions={mockActions}
          isEditor={false}
          modals={mockModals}
          auth={auth}
        />
      </Router>
    );

    const title = screen.getByTestId("Trabajos:");
    expect(title.innerHTML).toBe("Trabajos: 2");

    expect(container).toMatchSnapshot();
  });

  it("should render 5 book cards", () => {
    const { container } = render(
      <>
        <Router>
          <CardGrid
            type="book"
            actions={mockActions}
            isEditor={false}
            modals={mockModals}
            auth={auth}
          />
        </Router>
      </>
    );

    const title = screen.getByTestId("Obras:");
    expect(title.innerHTML).toBe("Obras: 5");

    expect(container).toMatchSnapshot();
  });
});
