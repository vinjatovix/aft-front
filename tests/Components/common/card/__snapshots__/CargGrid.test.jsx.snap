// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardGrid Component should render 2 work cards 1`] = `
<div>
  <div
    class="content-header"
  >
    <h1
      data-testid="Trabajos:"
    >
      Trabajos:
       
      2
    </h1>
  </div>
  <div
    class="card-grid"
  >
    <div
      class="card"
    >
      <ul
        class="work-list"
      >
        <li
          class=""
          data-testid="interprete"
        >
          by: 
          user
        </li>
        <li
          class=""
          data-testid="obra"
        >
          Obra: 
          Obra
        </li>
        <li
          class=""
          data-testid="personaje"
        >
          Personaje: 
          Personaje
        </li>
        <li
          class=""
          data-testid="escena"
        >
          Escena: 
          Escena
        </li>
        <li
          class=""
          data-testid="descripcion"
        >
          Descripción: 
          Descripción
        </li>
      </ul>
    </div>
    <div
      class="card"
    >
      <ul
        class="work-list"
      >
        <li
          class=""
          data-testid="interprete"
        >
          by: 
          user
        </li>
        <li
          class=""
          data-testid="obra"
        >
          Obra: 
          Obra
        </li>
        <li
          class=""
          data-testid="personaje"
        >
          Personaje: 
          Personaje
        </li>
        <li
          class=""
          data-testid="escena"
        >
          Escena: 
          Escena
        </li>
        <li
          class=""
          data-testid="descripcion"
        >
          Descripción: 
          Descripción
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`CardGrid Component should render 5 book cards 1`] = `
<div>
  <div
    class="content-header"
  >
    <h1
      data-testid="Obras:"
    >
      Obras:
       
      5
    </h1>
  </div>
  <div
    class="card-grid"
  >
    <div
      class="card"
    >
      <div
        class="flex-wrapper column"
      >
        <ul
          class="book"
          data-testid="book-detail"
        >
          <img
            alt="transmitter Refined"
            src="https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg"
          />
          <li
            class=""
            data-testid="Obra"
          >
            transmitter Refined
          </li>
          <li
            class="small"
            data-testid="Autor"
          >
            Theresa Larkin
          </li>
        </ul>
        <a
          class=""
          href="/book/6301216103b6fbb712bbde7c/characters"
        >
          <button
            class="bg-blue"
            data-testid="navigate-button"
          >
            🎭 Personajes
          </button>
        </a>
        <a
          class=""
          href="/book/6301216103b6fbb712bbde7c/scenes"
        >
          <button
            class="bg-blue"
            data-testid="navigate-button"
          >
            🎬 Escenas
          </button>
        </a>
      </div>
    </div>
    <div
      class="card"
    >
      <div
        class="flex-wrapper column"
      >
        <ul
          class="book"
          data-testid="book-detail"
        >
          <img
            alt="alarm Tasty"
            src="https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg"
          />
          <li
            class=""
            data-testid="Obra"
          >
            alarm Tasty
          </li>
          <li
            class="small"
            data-testid="Autor"
          >
            Jay Hintz
          </li>
        </ul>
        <a
          class=""
          href="/book/6301212c03b6fbb712bbde7a/characters"
        >
          <button
            class="bg-blue"
            data-testid="navigate-button"
          >
            🎭 Personajes
          </button>
        </a>
        <a
          class=""
          href="/book/6301212c03b6fbb712bbde7a/scenes"
        >
          <button
            class="bg-blue"
            data-testid="navigate-button"
          >
            🎬 Escenas
          </button>
        </a>
      </div>
    </div>
    <div
      class="card"
    >
      <div
        class="flex-wrapper column"
      >
        <ul
          class="book"
          data-testid="book-detail"
        >
          <img
            alt="Dr. microchip Small"
            src="https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg"
          />
          <li
            class=""
            data-testid="Obra"
          >
            Dr. microchip Small
          </li>
          <li
            class="small"
            data-testid="Autor"
          >
            Darrell Senger
          </li>
        </ul>
        <a
          class=""
          href="/book/62ffb64f42dc23efb3c3c0c4/characters"
        >
          <button
            class="bg-blue"
            data-testid="navigate-button"
          >
            🎭 Personajes
          </button>
        </a>
        <a
          class=""
          href="/book/62ffb64f42dc23efb3c3c0c4/scenes"
        >
          <button
            class="bg-blue"
            data-testid="navigate-button"
          >
            🎬 Escenas
          </button>
        </a>
      </div>
    </div>
    <div
      class="card"
    >
      <div
        class="flex-wrapper column"
      >
        <ul
          class="book"
          data-testid="book-detail"
        >
          <img
            alt="Mr. panel Practical"
            src="https://boxshot.com/3d-book-cover/how-to-make-a-3d-book-cover-in-photoshop/sample.jpg"
          />
          <li
            class=""
            data-testid="Obra"
          >
            Mr. panel Practical
          </li>
          <li
            class="small"
            data-testid="Autor"
          >
            Shaun Mohr
          </li>
        </ul>
        <a
          class=""
          href="/book/62ffb63c42dc23efb3c3c0c2/characters"
        >
          <button
            class="bg-blue"
            data-testid="navigate-button"
          >
            🎭 Personajes
          </button>
        </a>
        <a
          class=""
          href="/book/62ffb63c42dc23efb3c3c0c2/scenes"
        >
          <button
            class="bg-blue"
            data-testid="navigate-button"
          >
            🎬 Escenas
          </button>
        </a>
      </div>
    </div>
    <div
      class="card"
    >
      <div
        class="flex-wrapper column"
      >
        <ul
          class="book"
          data-testid="book-detail"
        >
          <img
            alt="El principito"
            src="https://media.v2.siweb.es/uploaded_thumb_medium/bd4e5fefd47a03f444d7dce77040dff3/principito_con_letra.jpg"
          />
          <li
            class=""
            data-testid="Obra"
          >
            El principito
          </li>
          <li
            class="small"
            data-testid="Autor"
          >
            Saint-Exupéry
          </li>
        </ul>
        <a
          class=""
          href="/book/62fed32c76d7b41305161d8a/characters"
        >
          <button
            class="bg-blue"
            data-testid="navigate-button"
          >
            🎭 Personajes
          </button>
        </a>
        <a
          class=""
          href="/book/62fed32c76d7b41305161d8a/scenes"
        >
          <button
            class="bg-blue"
            data-testid="navigate-button"
          >
            🎬 Escenas
          </button>
        </a>
      </div>
    </div>
  </div>
</div>
`;
